### PR TITLE
[KVStore] Add support for other data types

### PR DESCRIPTION
**Pull request by ZihengJiang**
@mli

### CODE REVIEW

**Comment by Godricly** [comment]
How do you support kvstore of DType when using ps-lite? If I remember correctly, the underling ps-lite is not supporting dtype.

**Comment by ZihengJiang** [comment]
@Godricly it only supports kvstore='local'/'device' for now

**Comment by ZihengJiang** [comment]
Discussion is needed due to this pr may involve some api change. The pr is raised to address dtype problem of kvstore, and it requires user to provide a `dtype` field in his DataIter: 
- https://github.com/ZihengJiang/mxnet/blob/kvstore/python/mxnet/model.py#L781
- https://github.com/ZihengJiang/mxnet/blob/kvstore/python/mxnet/model.py#L503

But looks like there exists some legacy case which uses tuple as `provide_data`? https://github.com/dmlc/mxnet/blob/master/tests/python/train/test_dtype.py#L130

@piiswrong @mli

**Comment by piiswrong** [comment]
model and module need to parse provide_data and convert to datadesc if necessary

**Comment by Godricly** [comment]
I tried 2 provide a infer datatype last year, LOL. #2559 #2564
What I suggest is to refactor ps-lite with better DType support before moving on in mxnet.[#58](https://github.com/dmlc/ps-lite/issues/58)

**Comment by mli** [comment]
do we need to update module as well?

**Comment by mli** [comment]
@Godricly ps-lite supports various data types. one can use templated SArray, such as SArray<int8>.  Internally ps-lite use char for communication. So any type with size >= sizeof(char) should be fine.

**Comment by ZihengJiang** [comment]
@mli module has been updated, it would be more appropriate to change the argument `data_shapes`/`label_shapes` to `data_attrs`/`label_attrs` here(https://github.com/ZihengJiang/mxnet/blob/kvstore/python/mxnet/module/module.py#L323), but I keep it for compatible.

File: python/mxnet/module/base_module.py [hunk 1]
```diff
@@ -64,16 +64,16 @@ def _check_names_match(data_names, data_shapes, name, throw):
             warnings.warn(msg)
 
 
-def _parse_data_desc(data_names, label_names, data_shapes, label_shapes):
```
Comment (piiswrong) [inline python/mxnet/module/base_module.py]: argument name changes has no practical benefit but can cause problems. Let's keep the original names and improve documents.

File: python/mxnet/module/module.py [hunk 2]
```diff
@@ -368,8 +368,10 @@ def bind(self, data_shapes, label_shapes=None, for_training=True,
             # that consumes the labels
             # assert label_shapes is not None
 
-        self._data_shapes, self._label_shapes = _parse_data_desc(
-            self.data_names, self.label_names, data_shapes, label_shapes)
+        data_attrs = data_shapes
+        label_attrs = label_shapes
+        self._data_attrs, self._label_attrs = _parse_data_desc(
```
Comment (piiswrong) [inline python/mxnet/module/module.py]: same here

**Pull request merged**